### ISSUE TITLE
nonce reuse issue

### ISSUE THREAD

**Issue by richiela** [ethereum/go-ethereum#15066]
#### System information

Geth version: 
Geth
Version: 1.6.7-stable
Architecture: amd64
Protocol Versions: [63 62]
Network Id: 1
Go Version: go1.8
Operating System: linux

CONFIG options:
```
geth  --fast --verbosity "3" --rpc --rpcaddr 0.0.0.0 --rpcport 8101 --rpccorsdomain "*" --rpcapi "db,net,eth,web3,personal,admin" --targetgaslimit 1000000 --cache 8192 --txpool.accountqueue 8192
```

OS & Version: Ubuntu 16.04
Commit hash : (if `develop`)

#### Expected behaviour
nonce's dont get reused

#### Actual behaviour
nonce was rolled back 300ish and re-used

#### Steps to reproduce the behaviour
We send out, at peak around 70-80tx's a min, and on avg around 40-50txs.  We use a gas price of 1.5x getGasPrice.  Whenever there is a spike in network traffic, and a quick increase in gasPrice, geth resets the nonce and overwrites entire blocks of transactions; in this case, about 300 nonces.  All sends are done via eth_sendTransaction json rpc.

Broadcasted: 2017-08-31 14:58:50.2500000 UTC -- This is reusing an old nonce.  This was an assigned nonce but hadnt confirmed yet.
```
> eth.getTransaction("0x3c6a1d4b327f62751ad2e6f677e16662a0ac3b31e6aef2b4789aa824a5dc86f2")
{
  blockHash: "0x0000000000000000000000000000000000000000000000000000000000000000",
  blockNumber: null,
  from: "0xfbb1b73c4f0bda4f67dca266ce6ef42f520fbb98",
  gas: 150334,
  gasPrice: 75000000000,
  hash: "0x3c6a1d4b327f62751ad2e6f677e16662a0ac3b31e6aef2b4789aa824a5dc86f2",
  input: "0x6ea056a900000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e5d0f042e878000",
  nonce: 1266479,
  r: "0x408ea73265466b227c8a1c8ae50a077ac62609a1f378ed68cc6707ed8a981647",
  s: "0x23ea18efc2a0dd32e65de252e75fddb8749592708fe8754456720cc900239046",
  to: "0x6e58d22a0a60225ce9bea0abe63648596247f394",
  transactionIndex: 0,
  v: "0x26",
  value: 0
}
```
Broadcasted: 2017-08-31 14:11:23.807 - this one confirms but not till Aug-31-2017 03:01:45 PM +UTC (according to etherscan)
```
> eth.getTransaction("0x73c9213e9c55b5e92d62d663fdd069c18a29c2ee3844f1e7ab4c08e7fdd140b2")
{
  blockHash: "0xc69ad32b41d4331a9396781171224375be32d53fe5ea6f13b4d68c5f436b8432",
  blockNumber: 4223742,
  from: "0xfbb1b73c4f0bda4f67dca266ce6ef42f520fbb98",
  gas: 121001,
  gasPrice: 90000000000,
  hash: "0x73c9213e9c55b5e92d62d663fdd069c18a29c2ee3844f1e7ab4c08e7fdd140b2",
  input: "0x",
  nonce: 1266479,
  r: "0x8f75d1fae36287fd595c5da641e75b9712dcbf2000dce56ad700ab6b3a7637dd",
  s: "0x36b3ae7e2a3fc60f95f24cd19489e993ab83a3e1355ccae311b83e57e0cb5c3",
  to: "0xfb793c7af5c0e6d8d1250a9c74ebadfaf3967fb6",
  transactionIndex: 4,
  v: "0x26",
  value: 374543630000000000
}
```
Another tx from the same rollback period:

Broadcasted: 2017-08-31 15:04:59.5430000
```
> eth.getTransaction("0x21c1956ad2b89326df750e550ce56cde25710550af2ec448b2520630856ff3f1")
{
  blockHash: "0x0000000000000000000000000000000000000000000000000000000000000000",
  blockNumber: null,
  from: "0xfbb1b73c4f0bda4f67dca266ce6ef42f520fbb98",
  gas: 150334,
  gasPrice: 75000000000,
  hash: "0x21c1956ad2b89326df750e550ce56cde25710550af2ec448b2520630856ff3f1",
  input: "0x6ea056a9000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001e5fcefaaf16c00",
  nonce: 1266540,
  r: "0x2f2f2f947a50f95617443a9da083ee762cddb3b03195650076800da9d522a9ee",
  s: "0x3291186846a3b09ff332b73ec31e77b61d139fe87148d3f8b4b4abc09853dee",
  to: "0x11fd1c5999acc7d5d83fd520f5b33b05807f1a85",
  transactionIndex: 0,
  v: "0x26",
  value: 0
}
```
Broadcasted: 2017-08-31 14:16:31.910 - not confirmed till Aug-31-2017 03:14:28 PM +UTC
```
> eth.getTransaction("0xe919fb38f553b65dc4bd2602460d99b88e6986cba5eb0e42d730f516a80f5d87")
{
  blockHash: "0x78dc81e0579eb65bfc4edb42ad20879b29c032a19f12c94f3da3fc6190f75999",
  blockNumber: 4223769,
  from: "0xfbb1b73c4f0bda4f67dca266ce6ef42f520fbb98",
  gas: 194064,
  gasPrice: 90000000000,
  hash: "0xe919fb38f553b65dc4bd2602460d99b88e6986cba5eb0e42d730f516a80f5d87",
  input: "0xa9059cbb000000000000000000000000df04eaf5fe642ab9fce3a9bb4957361f514bc6570000000000000000000000000000000000000000000000000000000f819c17c0",
  nonce: 1266540,
  r: "0x251be747647776f146c9bccd6cdc00cd6a9e83276c35e5ddf6072ece71a07b78",
  s: "0x1abd960edb35228248b4a9fe588ed5c2ce7672a795aff3cf23960132afa34d26",
  to: "0x419d0d8bdd9af5e606ae2232ed285aff190e711b",
  transactionIndex: 14,
  v: "0x25",
  value: 0
}


```
#### Backtrace

````
[backtrace]
````

**Comment by holiman** [comment]
>  a quick increase in gasPrice

I would guess that's it. The transactions laying in the pools are removed if they are under the gasPrice, so that could just drop 300 transactions. However, they _should_ be tagged as `local`s, and therefore retained anyway... cc @karalabe

**Comment by karalabe** [comment]
One thing we might need to take a look into are reorgs. If a block gets a transaction included, then a reorg happens, there might be a slight delay between reorging the "stale block" and getting the old transactions back into the tx pool. If a new transaction arrives in between, it will see the "next nonce" as low, as the previously included transactions haven't been rescheduled yet.

This is just a hunch btw, but I've seen some similar phenomenon on Rinkeby so it might be worth to investigate.

**Comment by holiman** [comment]
So, just to make this clear: are the two `eth.getTransaction` -invocations made at the same time? As in, did the client hold both these transactions simultaneously @richiela ?

**Comment by richiela** [comment]
Correct... all the above tx's were retrieved at the same time from the geth daemon.

**Comment by holiman** [comment]
@karalabe I'll use this ticket to collect paste any finding I make, and later collect them in a PR. First one: 

https://github.com/ethereum/go-ethereum/blob/master/core/tx_list.go#L437

* If `tx` is local, it won't be saved, since the loop will break before saving it. 
* Likely not the root cause here, because it would (afaict) still clean it from `all`, whereas in _this_ bug there's some flaw where a tx is left in `all` despite being removed from relevant pools.

**Comment by holiman** [comment]
I think this was auto-closed due to my PR text (?). I'm reopening, because I'm not certain it's fixed.

**Comment by karalabe** [comment]
@richiela We've found and fixed an ugly issue in the transaction pool. Hopefully that will make nonces a lot more reliable from now on. It will be part of the 1.7.0 release.

**Comment by richiela** [comment]
Looking forward to it... do we have a timeline for 1.7.0?  You are correct, unstable builds is no beuno

**Comment by karalabe** [comment]
@richiela We've released 1.7, but please take care with it, since it contains a significant number of updates and it's not possible to downgrade to 1.6.x from 1.7.

The error we found was that during reorgs (uncles, concurrent blocks, etc), the system fired two events: "new head block" and "reorged transactions". The pool updated its state (i.e. account nonces) based on the first event, and readded removed transactions based on the second event. But events are fired concurrently, so if the tx event arrived first, all txs were discarded as conflicting with the nonces tracked by the pool; and only afterwards were the nonces updated, when the head block event arrived.

There was a similar issue around mining too btw, as the miner also did some funky crap concurrently with the pool. In 1.7.0, we've reworked the pool to only ever have a single "modification" pathway and all updates to happen atomically, instead reactively to various system events.

**Comment by holiman** [comment]
I suspect this is fixed now, closing
